add: fast data loading section (#1391)
mia-platform/documentation: docs/fast_data/concepts/data_loading.mdx

---
id: data_loading
title: Fast Data Loading
sidebar_label: Data Loading
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> __Data Loading__ is a fundamental process to introduce information from the Systems of Record:

Reviewer: ```suggestion
__Data Loading__ is a fundamental process to introduce information from the Systems of Record to your Fast Data system:
```

:::tip
If you need to Bootstrap either Fast Data services or infrastructural component, you can check out [this section](/fast_data/faq/bootstrapping_fast_data.md#what-should-be-the-required-resources-to-perform-an-initial-load). 
:::

Before an Initial Load takes place, you have to prepare your [Fast Data architecture](/fast_data/concepts/architecture.md#putting-it-all-together) accordingly.

<Tabs groupId='fast-data-architecture'>

<TabItem value="event-driven" label="Event-Driven Architecture" default>

First, you have to check that ingestion topics have been correctly defined for the Projections that needs to be populated.

Then, if there exists a Single View Trigger Generator that consumes the messages produced into the `pr-update` topics of
the projections, you have to check that message consuming is not up-and-running (either by using Runtime Management or setting replicas to zero).

In this way, during the ingestion phase, strategies will not be executed. Once the ingestion procedure has been completed, you can restart
the SVTG, so that strategies will generate the single view identifiers accordingly.

##### Using Runtime Management

> With Runtime Management, is possible to manage Initial Load operations without altering your configuration.

Reviewer: We may put a link to runtime management section

Author: link added [here](https://github.com/mia-platform/documentation/pull/1391/commits/730e00d01bbf09241b3c687c05224456843836a7)

> From the Control Plane frontend, you have to check first that the Domain Pipeline that will consume messages

Reviewer: Alternatively, here we may put a link to runtime management fast data section with control plane info

Author: link added [here](https://github.com/mia-platform/documentation/pull/1391/commits/730e00d01bbf09241b3c687c05224456843836a7)

> is in a running state. Then, in the Single View pipeline that will receives data streams

Reviewer: ```suggestion
is in a running state. Then, in the Single View pipeline that will receive data streams
```

from the projections involved in the Initial Load, both `Collected` and `Curated` stages must be in a __Paused State__.

The state of your pipeline should match the example shown below, where a data stream generating a single view 
named `orders-history` is completely paused.  

![Initial Load Runtime Management](img/data_loading_initial_load_runtime.png)

Then, once the Initial Load has been completed and messages have been stored correctly on MongoDB, you can put the `Collected`
stage back to the __Running State__ and let the underlying SVTG consuming the `pr-update` messages to generate the `trigger` messages.

![Initial Load Runtime Management](img/data_loading_initial_load_runtime_restart_collected.png)

:::tip

> If the Initial Load involves projections having large datasets, is suggested to __only__ restart the base projection, i.e. the projection

Reviewer: may have a link to base projection documentation details?

Author: link added [here](https://github.com/mia-platform/documentation/pull/1391/commits/730e00d01bbf09241b3c687c05224456843836a7)

</TabItem>

</Tabs>

## Full Refresh

In Fast Data, __Full Refresh__ operations are made to regenerate the documents of a whole single view collection.

### Overview

The operation is carried out within the Fast Data Architecture, to let the Single View Creator (SVC) performing
the aggregation of all the documents of a Single View collection.  

### Scenarios 

This operation may be needed to make the single view aggregation consistent with the source system when:

- fields have been _added_/_removed_ _to_/_from_ a projection after the Initial Load and 
  needs to be _added_/_removed_ _into_/_from_ the Single View;
- there has been some _disalignment_ in the single view generation process, due to either
  infrastractural problems or misconfigurations.

### Best Practices

Before refreshing a single view collection, be sure that strategies have been disabled:
this is to ensure that projections are not changing while the refresh operation is taking place,
which benefits also the memory usage of your database infrastructure.

Below you can find the instructions to setup your runtime based on the

> [Fast Data architecture](/fast_data/concepts/architecture.md#putting-it-all-together) you've deployed.

Reviewer: ```suggestion
[Fast Data architecture](/fast_data/concepts/architecture.md#putting-it-all-together) you have deployed.
```
